File: problems/Swift/506.相对名次.swift
/**
 * 506.相对名次 [Easy]
 * Time Complexity: O(n*log(n)), Space Complexity: O(n)
 * https://leetcode-cn.com/problems/relative-ranks/
 */

class Solution {
  func findRelativeRanks(_ score: [Int]) -> [String] {
    let desc = ["Gold Medal", "Silver Medal", "Bronze Medal"]
    var arr: [[Int]] = []

    for i in 0 ..< score.count {
      arr.append([score[i], i])
    }
    arr.sort(by: { $0[0] > $1[0] })

    var ans = [String](repeating: "", count: score.count)
    for i in 0 ..< score.count {
      if i < 3 {
        ans[arr[i][1]] = desc[i]
      } else {
        ans[arr[i][1]] = "\(i + 1)"
      }
    }

    return ans
  }
}

File: problems/rust-solutions/src/506.相对名次.rs
// 506.相对名次 [Easy]
// Time Complexity: O(n*log(n)), Space Complexity: O(n)
// https://leetcode-cn.com/problems/relative-ranks/
use std::cmp::Reverse;

impl Solution {
  pub fn find_relative_ranks(score: Vec<i32>) -> Vec<String> {
    let count = score.len();
    let mut arr: Vec<(usize, i32)> = score.into_iter().enumerate().collect();
    let mut ans: Vec<String> = (0..count).map(|_| String::new()).collect();
    arr.sort_unstable_by_key(|x| Reverse(x.1));
    arr.into_iter().enumerate().for_each(|(i, (idx, _))| {
      ans[idx] = Solution::get_name(i + 1);
    });
    return ans;
  }

  fn get_name(x: usize) -> String {
    match x {
      1 => String::from("Gold Medal"),
      2 => String::from("Silver Medal"),
      3 => String::from("Bronze Medal"),
      _ => x.to_string(),
    }
  }
}
